Prometheus metrics for connection tracking garbage

eloycoto opened the pull request:
Hi, 

 Added the following metrics to the system:
                                           
 cilium_datapath_conntrack_gc_duration     
 cilium_datapath_conntrack_gc_expired_total
 cilium_datapath_conntrack_gc_key_fallbacks
 cilium_datapath_conntrack_gc_runs_total   
 cilium_datapath_conntrack_gc_size         
 cilium_datapath_errors_total              

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5296)
<!-- Reviewable:end -->

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -2664,23 +2664,37 @@ func (e *Endpoint) doGarbageCollectConntrack(isIPv6 bool, filter *ctmap.GCFilter
+		deletedIpv4 = e.doGarbageCollectConntrack(false, filter)
+		logFields["Ipv4DeletedConnections"] = deletedIpv4
 	}
+	e.getLogger().WithFields(logFields).Debug("Deleted connections from Conntrack table")
```
Why is this needed? We already have an extensive log line in `doGC*()` via `finish()`:
```
        log.WithFields(logrus.Fields{
                logfields.StartTime: s.started,
                logfields.Duration:  s.finished.Sub(s.started),
                "numDeleted":        s.deleted,
                "numLookups":        s.count,
                "numLookupsFailed":  s.lookupFailed,
                "numKeyFallbacks":   s.keyFallback,
                "completed":         s.completed,
                "maxEntries":        s.maxEntries,
        }).Infof("%s Conntrack garbage collection statistics", s.family)
```

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -2664,23 +2664,37 @@ func (e *Endpoint) doGarbageCollectConntrack(isIPv6 bool, filter *ctmap.GCFilter
+// RunConntrackGC runs the conntrack garbace collector for the endpoint and the
+// given filter.
+func (e *Endpoint) RunConntrackGC(filter *ctmap.GCFilter) {
+	e.UnconditionalLock()
```
This lock will be horrible on performance. We can't take it like this. Why is it needed?

tgraf commented on pkg/endpointmanager/conntrack.go:
```diff
@@ -71,72 +71,66 @@ func runGC(e *endpoint.Endpoint, isIPv6 bool, filter *ctmap.GCFilter) {
+// endpoint that has been restored. No new endpoints can appear yet so we can
+// assume that any other entry not belonging to a restored endpoint has become
+// stale.
+func scrubAllNonRestoredIPs(ipv6 bool, restoredEndpoints []*endpoint.Endpoint) {
```
This description is wrong. The function is called on every gc run and is responsible to do GC on per endpoint CT tables. The condition is: `len(eps) > 0 || initialScan`

tgraf commented on pkg/metrics/metrics.go:
```diff
@@ -118,6 +119,14 @@ var (
+	// ConntrackGC is the number of deleted entries on the conntrack table by
+	// the garbage collector process.
+	ConntrackGC = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: Namespace,
```
Subsystem: Datapath,?

tgraf commented on pkg/metrics/metrics.go:
```diff
@@ -118,6 +119,14 @@ var (
+	// the garbage collector process.
+	ConntrackGC = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: Namespace,
+		Name:      "conntrack_gc",
```
conntrack_gc_expired_total

tgraf commented on pkg/metrics/metrics.go:
```diff
@@ -118,6 +119,14 @@ var (
 
+	// ConntrackGC is the number of deleted entries on the conntrack table by
+	// the garbage collector process.
+	ConntrackGC = prometheus.NewGaugeVec(prometheus.GaugeOpts{
```
This should be a CounterVec not a GaugeVec

eloycoto replied:
Hi!

Are you sure that should be a counter? I don't think so, Prometheus doc says that the counter is a cumulative information about a metric, so will never be 0. In the graph, you'll see that the number only increase, but you don't see how many connections were deleted. 

If you have a gauge, you can see how many connections are dropping every interval, plus if one endpoint deletes a lot in a time, you can set an alarm, if it's a count it's super hard to set alarms. 

What do you think?
Regards

tgraf replied:
Yes, I think it should be a counter. What you describe above is the rate and the rate can be extracted from a counter:
```
rate(conntrack_gc_expired_total{table="globale"}[5m])
```

tgraf commented on pkg/endpointmanager/conntrack.go:
```diff
@@ -69,6 +71,8 @@ func runGC(e *endpoint.Endpoint, isIPv6 bool, filter *ctmap.GCFilter) {
 		}).Debug("Deleted filtered entries from map")
 	}
+
+	metrics.ConntrackGC.WithLabelValues(metricLabel).Set(float64(deleted))
```
Needs a label to indicate whether it is ipv4 or ipv6

tgraf commented on pkg/metrics/metrics.go:
```diff
@@ -118,6 +119,14 @@ var (
+		Name:      "conntrack_gc",
+		Help:      "Number of deleted entries in the conntrack table by the garbage collector process",
+	}, []string{logfields.EndpointID})
+
```
Let's also add CounterVec for:

* conntrack_gc_runs_total - ++ whenever a gc is run
* conntrack_gc_incomplete_runs_total - ++ when gcStats.completed == false
* conntrack_gc_key_fallbacks - gcStats.keyFallback

GaugeVec:
* conntrack_gc_size - gcStats.count *IF* gcStats.completed

Timer:
* conntrack_gc_duration = gcStats.finished.Sub(gcStats.started)

eloycoto committed:
Metrics: Add metrics for the conntrack garbage collector

Added the following metrics to the system:

cilium_datapath_conntrack_gc_duration_seconds
cilium_datapath_conntrack_gc_key_fallbacks_total
cilium_datapath_conntrack_gc_runs_total
cilium_datapath_conntrack_gc_entries

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

raybejjani commented on Documentation/configuration/metrics.rst:
```diff
@@ -42,6 +42,12 @@ Datapath
+* ``datapath_conntrack_gc_runs_total``: Number of times that the conntract garbage collector process was run
+* ``datapath_conntrack_gc_incomplete_runs_total``: Number of times that the conntrack garbage collector failed
+* ``datapath_conntrack_gc_key_fallbacks``: Number of times that the Key fallback was invalid.
+* ``datapath_conntrack_gc_size``: The size of the conntrack table when the garbage collector ran
```
It might be worth mentioning if this was at the start or end of the GC run

raybejjani commented on Documentation/configuration/metrics.rst:
```diff
@@ -42,6 +42,12 @@ Datapath
+* ``datapath_conntrack_gc_incomplete_runs_total``: Number of times that the conntrack garbage collector failed
+* ``datapath_conntrack_gc_key_fallbacks``: Number of times that the Key fallback was invalid.
+* ``datapath_conntrack_gc_size``: The size of the conntrack table when the garbage collector ran
+* ``datapath_conntrack_gc_duration``: Duration in milliseconds of the garbage collector process
```
I'm afraid prometheus is picky about this. It wants all times in seconds. Can we convert to seconds to avoid any confusion?

raybejjani commented on Documentation/configuration/metrics.rst:
```diff
@@ -42,6 +42,12 @@ Datapath
+* ``datapath_conntrack_gc_expired_total``: Number of deleted entries in the conntrack table by the garbage collector process
+* ``datapath_conntrack_gc_runs_total``: Number of times that the conntract garbage collector process was run
+* ``datapath_conntrack_gc_incomplete_runs_total``: Number of times that the conntrack garbage collector failed
+* ``datapath_conntrack_gc_key_fallbacks``: Number of times that the Key fallback was invalid.
```
Is this the number of times we had 1 bad key and used the fallback? or when that fallback was also bad?
If it's the second one, does that trigger a restart of the GC? Would it be better to record conntrack_gc_restart_total and then us a label for the reason? This would only be useful if we have other reasons to restart, or if we wanted to establish a convention around metrics like this for other GC/cleanup things.

tgraf replied:
This is the number of successful key fallbacks. Any successful or unsuccessful GC run may have 0..n key fallbacks. The interrupted counter is for when we restart

raybejjani commented on Documentation/configuration/metrics.rst:
```diff
@@ -42,6 +42,12 @@ Datapath
 --------
 
 * ``datapath_errors_total``: Total number of errors occurred in datapath management, labeled by area, name and address family.
+* ``datapath_conntrack_gc_expired_total``: Number of deleted entries in the conntrack table by the garbage collector process
```
Are there any cases where we might delete an entry that isn't expired? An example would be if we are cleaning up entries for a non-existent endpoint? If so, perhaps calling this `datapath_conntrack_gc_deleted_total` and then using a label for the reason would make it easier to split out the various scenarios.

tgraf replied:
Good point. I'm fine with this as well

raybejjani commented on pkg/metrics/metrics.go:
```diff
@@ -253,8 +259,49 @@ var (
+		Namespace: Namespace,
+		Subsystem: Datapath,
+		Name:      "conntrack_gc_deleted",
+		Help:      "Number of deleted entries in the conntrack table by the garbage collector process",
```
Could we change this to: "Number of garbage collected conntrack table entries"

raybejjani commented on pkg/metrics/metrics.go:
```diff
@@ -253,8 +259,49 @@ var (
+		Namespace: Namespace,
+		Subsystem: Datapath,
+		Name:      "conntrack_gc_runs_total",
+		Help:      "Number of times that the conntract garbage collector process was run",
```
conntract is typoed.

raybejjani commented on pkg/metrics/metrics.go:
```diff
@@ -253,8 +259,49 @@ var (
+		Namespace: Namespace,
+		Subsystem: Datapath,
+		Name:      "conntrack_gc_size",
+		Help:      "The size of the conntrack table when the garbage collector ran",
```
Could we be even more specific here? "The number of entries in the conntrack table at the *start* of a GC run"

raybejjani replied:
Oh, I noticed that gcStats has an aliveEntries field. Can we use that with label here? or is it guaranteed that aliveEntries is the final number left in the table?

Also, I don't see where `ConntrackGCSize` is set

raybejjani commented on pkg/metrics/metrics.go:
```diff
@@ -253,8 +259,49 @@ var (
+		Namespace: Namespace,
+		Subsystem: Datapath,
+		Name:      "conntrack_gc_duration",
+		Help:      "Duration in milliseconds of the garbage collector process",
```
milliseconds -> seconds

raybejjani commented on pkg/metrics/metrics.go:
```diff
@@ -253,8 +259,49 @@ var (
+		Namespace: Namespace,
+		Subsystem: Datapath,
+		Name:      "conntrack_gc_key_fallbacks",
+		Help:      "Number of times that the Key fallback was invalid.",
```
This was still unclear to me, in light of what tgraf said. Could we change it to "Number of times a key fallback was needed when iterating over the BPF map"

raybejjani reviewed (changes requested):
I think we missed setting `ConntrackGCSize`, that's the only real blocker

tgraf commented on Documentation/configuration/metrics.rst:
```diff
@@ -41,7 +41,20 @@ Endpoint
-* ``datapath_errors_total``: Total number of errors occurred in datapath management, labeled by area, name and address family.
+* ``datapath_errors_total``: Total number of errors occurred in datapath
+  management, labeled by area, name and address family.
+* ``datapath_conntrack_gc_deleted``: Number of deleted entries in the
```
Given this is a counter, it should have the _total suffix

eloycoto replied:
This is a old comment, now it's a gauge.

joestringer replied:
It should still have units suffix, for example "_count"

tgraf commented on Documentation/configuration/metrics.rst:
```diff
@@ -41,7 +41,20 @@ Endpoint
+* ``datapath_conntrack_gc_deleted``: Number of deleted entries in the
+  conntrack table by the garbage collector process. It has a scope label that
+  provides information with the reason why the connection was deleted.
+* ``datapath_conntrack_gc_runs``: Number of times that the conntrack garbage
```
Given this is a counter, it should have _total suffixes

tgraf commented on Documentation/configuration/metrics.rst:
```diff
@@ -41,7 +41,20 @@ Endpoint
+* ``datapath_conntrack_gc_runs``: Number of times that the conntrack garbage
+  collector process was run. It contains a label completed that describes if it
+  was successful or not.
+* ``datapath_conntrack_gc_key_fallbacks``: Number of times that the Key fallback
```
Given this is a counter, it should have _total suffixes

tgraf commented on Documentation/configuration/metrics.rst:
```diff
@@ -41,7 +41,20 @@ Endpoint
+  was invalid.
+* ``datapath_conntrack_gc_size``: The size of the conntrack table when the
+  garbage collector started to run
+* ``datapath_conntrack_gc_duration``: Duration in seconds of the garbage
```
This should have the suffix _seconds

tgraf reviewed (changes requested):
It's typically helpful to run:
https://github.com/prometheus/prometheus/blob/master/util/promlint/promlint.go

tgraf commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -349,22 +350,30 @@ func statStartGc(m *bpf.Map, family gcFamily) gcStats {
 	}
 
-	if !s.completed {
+	metrics.ConntrackGCRuns.WithLabelValues(family, strconv.FormatBool(s.completed)).Inc()
```
A label of just "true" or "false" is a bit weird to track completion. Can we add "completed=true" and "completed=false"?

eloycoto replied:
what about label name `status` and values `completed` and `incompleted` make more sense?

tgraf replied:
That's fine

joestringer replied:
Usually it's `uncompleted` or `incomplete`, rather than `incompleted`.

tgraf commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -349,22 +350,30 @@ func statStartGc(m *bpf.Map, family gcFamily) gcStats {
+	metrics.ConntrackGCRuns.WithLabelValues(family, strconv.FormatBool(s.completed)).Inc()
+
+	if s.completed {
+		metrics.ConntrackGCSize.WithLabelValues(family).Add(float64(s.count))
```
This needs to use `Set()`

I was wrong, this will need to be different, `count` is the number of lookups performed. It can be > than the total size if we got interrupted. This must be set to `aliveEntries`

tgraf commented on pkg/metrics/metrics.go:
```diff
@@ -253,8 +259,49 @@ var (
+
+	// ConntrackGCDeleted is the number of deleted entries in the
+	// conntrack table that has expired.
+	ConntrackGCDeleted = prometheus.NewCounterVec(prometheus.CounterOpts{
```
I take back my previous feedback. It makes sense to make this a Gauge and combine it with `ConntrackGCSize` and add a label which indicates whether the entry is `alive` or `dead` so we can have a graph which shows the number of entries as a line and the history of deleted entries in relation to that.

Sorry about the extra work caused by my feedback.

raybejjani commented on pkg/endpointmanager/conntrack.go:
```diff
@@ -39,7 +49,7 @@ const (
 // garbage collected and any failures will be logged to the endpoint log.
 // Otherwise it will garbage-collect the global map and use the global log.
-func runGC(e *endpoint.Endpoint, isIPv6 bool, filter *ctmap.GCFilter) {
+func runGC(e *endpoint.Endpoint, isIPv6 bool, filter *ctmap.GCFilter) int {
```
Could you update the comment to say what's returned? or maybe use a named return value?

raybejjani commented on pkg/metrics/metrics.go:
```diff
@@ -253,8 +259,49 @@ var (
+	}, []string{LabelDatapathArea, LabelDatapathName, LabelDatapathFamily})
+
+	// ConntrackGCDeleted is the number of deleted entries in the
+	// conntrack table that has expired.
```
"that _have_ expired"

raybejjani commented on pkg/metrics/metrics.go:
```diff
@@ -253,8 +259,49 @@ var (
+		Namespace: Namespace,
+		Subsystem: Datapath,
+		Name:      "conntrack_gc_deleted",
+		Help:      "Number of garbage collected conntrack table entries",
```
It would be good to specify that this is per-run.
"Number of deleted conntrack table entries in a single GC run." or something

raybejjani commented on pkg/metrics/metrics.go:
```diff
@@ -253,8 +259,49 @@ var (
+		Subsystem: Datapath,
+		Name:      "conntrack_gc_deleted",
+		Help:      "Number of garbage collected conntrack table entries",
+	}, []string{LabelDatapathFamily, LabelScope})
```
Do need to track ipv4 vs ipv6 here? or is that the scope? In either case, the Help should mention that since it will be important.

joestringer replied:
That should be LabelDatapathFamily.

raybejjani commented on pkg/metrics/metrics.go:
```diff
@@ -253,8 +259,49 @@ var (
+		Namespace: Namespace,
+		Subsystem: Datapath,
+		Name:      "conntrack_gc_runs_total",
+		Help:      "Number of times that the conntrack garbage collector process was run",
```
perhaps mention: "labelled by completion status"

raybejjani commented on pkg/metrics/metrics.go:
```diff
@@ -253,8 +259,49 @@ var (
+		Subsystem: Datapath,
+		Name:      "conntrack_gc_duration_seconds",
+		Help:      "Duration in seconds of the garbage collector process",
+	}, []string{LabelDatapathFamily})
```
We need to track whether this was ipv4 vs ipv6, and if the GC run completed or not, don't we? otherwise the data isn't as useful.

joestringer replied:
Again, DatapathFamily is a protocol family, ie IPv4 or IPv6.

raybejjani commented on pkg/metrics/metrics.go:
```diff
@@ -253,8 +259,49 @@ var (
+
+	// ConntrackGCDeleted is the number of deleted entries in the
+	// conntrack table that has expired.
+	ConntrackGCDeleted = prometheus.NewGaugeVec(prometheus.GaugeOpts{
```
I thought we were combining this gauge with `ConntrackGCSize`? The combined gauge could be updated at the end of the GC run with:
`.Set(deleted)` with family=ipv4/ipv6, status=deleted
`.Set(alive)` with family=ipv4/6, status=alive

This way making the graph is simpler to see:
- how many conntrack entries do we have
- how many conntrack entries for ipv4/6 do we have
- Is the conntrack table filling up

joestringer commented on pkg/endpointmanager/conntrack.go:
```diff
@@ -29,6 +29,16 @@ import (
 	MinGcInterval int = 5
+
+	// Global string for refering in metrics
+	Global string = "global"
```
Is this still used?
Does this need to be exported? 
What type of label is this? Is this a scope? In that case it would be helpful if at least the comment described this, for example:

`Global determines the scope of conntrack metrics to be for an entire garbage collection round of the the global, shared conntrack map.`

eloycoto replied:
sorry, didn't push without that :-(

joestringer commented on pkg/endpointmanager/conntrack.go:
```diff
@@ -29,6 +29,16 @@ import (
+	Global string = "global"
+
+	// Expires string to refering on metrics
+	expires string = "expires"
```
Similarly, the comment would be useful if it includes what it actually means (ie it's a metrics value for a scope, and how that scope measures part of the garbage collection).

eloycoto replied:
sorry, didn't push without that :-(

joestringer commented on pkg/endpointmanager/conntrack.go:
```diff
@@ -29,6 +29,16 @@ import (
+
+	// CleanOnStart string to refering that the CT was clean because it was
+	// cleaning the restored endpoints CT.
+	CleanOnStart string = "cleanOnStart"
```
Does this need to be exported?

It would also make this one clearer if the comment states that it's a scope.

eloycoto replied:
sorry, didn't push without that :-(

joestringer commented on pkg/metrics/metrics.go:
```diff
@@ -71,6 +71,12 @@ var (
+	// LabelStatus the label from completed task
+	LabelStatus = "status"
+
+	// LabelScope is the label used to defined multiples scopes in the same metric
```
Perhaps, add "For example, one counter may measure a metric over the scope of the entire event (scope=global), or just part of an event (scope=slow_path)"?

joestringer commented on Documentation/configuration/metrics.rst:
```diff
@@ -41,7 +41,18 @@ Endpoint
+* ``datapath_conntrack_gc_runs_total``: Number of times that the conntrack
+  garbage collector process was run. It contains a label status that describes
+  if it was successful or not.
+* ``datapath_conntrack_gc_key_fallbacks_total``: Number of times that the Key fallback
```
I'm wondering since this is a "per-entry" type statistic (may occur multiple times per GC run), perhaps it also makes sense to track in the same metric as "datapath_conntrack_gc_size" ?

eloycoto replied:
@raybejjani ^^^^

tgraf replied:
Discussed this briefly yesterday and the worry was mostly that the scale of key_fallbacks_total is very different from the total size. The total size may be 200K while the keyfallbacks maybe 0..10

tgraf replied:
We could clearly stuff everything into a single metric. I'm not sure what the best practice is to be honest. My gut feeling tells me not to mix counters and gauges that have a completely different scale. That said, we can have multiple graphs based on a single gauge each grouped by certain labels. I'm not sure it really matters that much. Most people will use the standard dashboard.

joestringer commented on Documentation/configuration/metrics.rst:
```diff
@@ -41,7 +41,18 @@ Endpoint
+  if it was successful or not.
+* ``datapath_conntrack_gc_key_fallbacks_total``: Number of times that the Key fallback
+  was invalid.
+* ``datapath_conntrack_gc_size``: The size of the conntrack table when the
```
I think it's a "_count" now, rather than a "_size"? Since it includes deleted?

(I realise it's a gauge, but this just means it's an instantaneous measurement, which I think still uses individual counts as the unit)

tgraf replied:
`_count` should never be used except for histograms.
https://github.com/prometheus/prometheus/blob/master/util/promlint/promlint.go#L185

joestringer reviewed (commented):
I have a few suggestions, apologies if I'm a bit late to the party.

In particular, I wonder if the keyFallbacks should be still its own metric or if it should be folded into the same place as alive/deleted entries. Really I see that metric as measuring what the `gcStats` structure measures.

joestringer commented on pkg/endpointmanager/conntrack.go:
```diff
@@ -69,6 +69,7 @@ func runGC(e *endpoint.Endpoint, isIPv6 bool, filter *ctmap.GCFilter) {
 			"count":        deleted,
 		}).Debug("Deleted filtered entries from map")
 	}
+	return
```
This seems extraneous?

joestringer commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -349,22 +352,33 @@ func statStartGc(m *bpf.Map, family gcFamily) gcStats {
+		metrics.ConntrackGCSize.WithLabelValues(family, metricsAlive).Set(float64(s.aliveEntries))
+		metrics.ConntrackGCSize.WithLabelValues(family, metricsDeleted).Set(float64(s.deleted))
+	} else {
+		status = "uncompleted"
```
I think we may get statistics even in the incomplete case here, maybe we should also increment them in this other path?

joestringer replied:
The caveat would be that if the dump is incomplete, maybe the statistics aren't entirely correct (though reporting what's known could still be useful).

eloycoto replied:
No, a previous comment said that we shouldn't report if it's not successful. The  alive can be 0 if it is not completed, so I understand that we don't want to report that 0 entries in the CT. 

The other thing is to use s.Count, the number when start, but a previous comment also said that we should use alive. 

This is the best way that I found, sorry :-(

joestringer reviewed (approved):
Looks good, maybe we'll miss some statistics if the dump doesn't complete though.

tgraf commented on pkg/metrics/metrics.go:
```diff
@@ -71,6 +71,15 @@ var (
+	// metric.For example, one counter may measure a metric over the scope of
+	// the entire event (scope=global), or just part of an event
+	// (scope=slow_path)
+	LabelScope = "scope"
```
Where is this used?

eloycoto replied:
sorry, was used in another metrics and I forget to deleted.

tgraf commented on pkg/metrics/metrics.go:
```diff
@@ -253,8 +262,43 @@ var (
+		Namespace: Namespace,
+		Subsystem: Datapath,
+		Name:      "conntrack_gc_entries",
+		Help: "The number of entries in the conntrack table at the end of a GC run " +
```
You are mixing spelling out `garbage collector` and `GC`. I would rewrite this as:

"The number of alive and deleted conntrack entries at the end of a garbage collector run labeled by datapath family."

raybejjani commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -349,22 +352,33 @@ func statStartGc(m *bpf.Map, family gcFamily) gcStats {
+		metrics.ConntrackGCSize.WithLabelValues(family, metricsAlive).Set(float64(s.aliveEntries))
+		metrics.ConntrackGCSize.WithLabelValues(family, metricsDeleted).Set(float64(s.deleted))
+	} else {
+		status = "uncompleted"
```
I'll aplogise up front: I'm sorry.

We don't seem to use uncompleted anywhere else, but we do use incomplete. Since the label for this will end up used in other describe-a-run metrics (I hope, at least), would anyone object to having status={complete, incomplete} ? This isn't a big deal to me, so if we just want to merge and move on that's ok too.

eloycoto replied:
https://github.com/cilium/cilium/pull/5296#discussion_r213336386

joestringer replied:
TBH when I look again, "uncompleted" looks a bit weird to me, but I don't really want to be a pain and ask to change it again ;)

raybejjani commented on Documentation/configuration/metrics.rst:
```diff
@@ -41,7 +41,17 @@ Endpoint
+* ``datapath_conntrack_gc_entries``: The number of alive and deleted conntrack
+  entries at the end of a garbage collector run labeled by datapath family.
+* ``datapath_conntrack_gc_duration_seconds``: Duration in seconds of the garbage
+  collector process labelled by datapath and completation status.
```
This is my fault 😭 labelled should be labeled to match US spelling, and the other descriptions we have. Sorry.

eloycoto replied:
done
